pkg/route/apiserver/admission/requiredrouteannotations/admission.go

```diff
@@ -0,0 +1,337 @@
+	if a.GetResource().GroupResource() != routev1.Resource("routes") {
+		return nil
+	}
+	if _, isRoute := a.GetObject().(*routeapi.Route); !isRoute {
+		return nil
+	}
+
+	// Wait for all caches to sync (TODO: we could do this one by one for each cache too)
+	if !o.waitForSyncedStore(ctx) {
+		return admission.NewForbidden(a, errors.New(pluginName+": caches not synchronized"))
+	}
+
+	ingress, err := o.ingressLister.Get("cluster")
+	if err != nil {
+		return admission.NewForbidden(a, err)
+	}
+
+	newRoute := a.GetObject().(*routeapi.Route)
+	namespace, err := o.nsLister.Get(newRoute.Namespace)
+	if err != nil {
+		return admission.NewForbidden(a, err)
+	}
+
+	if err = isRouteHSTSAllowed(ingress, newRoute, namespace); err != nil {
+		return admission.NewForbidden(a, err)
+	}
+	return nil
+}
+
+/*
+func (o *requiredRouteAnnotations) SetRESTClientConfig(restClientConfig rest.Config) {
+	var err error
+
+	configClient, err := configv1client.NewForConfig(&restClientConfig)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+
+	// Use an informer for config.  Get the informer factory.
+	configInformers := configv1informers.NewSharedInformerFactory(configClient, timeToWaitForCacheSync)
+
+	// Set the flag if the informer has synced
+	o.ingressListerSynced = configInformers.Config().V1().Ingresses().Informer().HasSynced
+
+	// Set the ingresses Lister from the factory
+	o.ingressLister = configInformers.Config().V1().Ingresses().Lister()
+
+	// Start the informer
+	stopChan := make(chan struct{})
+	defer close(stopChan)
+	go configInformers.Start(stopChan)
+
+	o.routeClient, err = routetypedclient.NewForConfig(&restClientConfig)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+}
+*/
+
+func (o *requiredRouteAnnotations) SetExternalKubeInformerFactory(kubeInformers informers.SharedInformerFactory) {
+	o.nsLister = kubeInformers.Core().V1().Namespaces().Lister()
+	//o.nsListerSynced = kubeInformers.Core().V1().Namespaces().Informer().HasSynced
+	o.cachesToSync = append(o.cachesToSync, kubeInformers.Core().V1().Namespaces().Informer().HasSynced)
+}
+
+func (o *requiredRouteAnnotations) waitForSyncedStore(ctx context.Context) bool {
+	syncCtx, cancelFn := context.WithTimeout(ctx, timeToWaitForCacheSync)
+	defer cancelFn()
+	return cache.WaitForNamedCacheSync("RequiredRouteAnnotationsAdmissionPlugin", syncCtx.Done(), o.cachesToSync...)
+}
+
+func (o *requiredRouteAnnotations) ValidateInitialization() error {
+	if o.ingressLister == nil {
+		return fmt.Errorf(pluginName + " plugin needs an ingress lister")
+	}
+	if o.routeLister == nil {
+		return fmt.Errorf(pluginName + " plugin needs a route lister")
+	}
+	if o.nsLister == nil {
+		return fmt.Errorf(pluginName + " plugin needs a namespace lister")
+	}
+	if len(o.cachesToSync) < 3 {
+		return fmt.Errorf(pluginName + " plugin missing informer synced functions")
+	}
+	return nil
+}
+
+func NewRequiredRouteAnnotations() (admission.Interface, error) {
```

Comment:
```suggestion
func NewRequiredRouteAnnotations() (*requiredRouteAnnotations, error) {
```
It's better to return the concrete type here.